From 97f7bc2fa3f5dabc9be64e01d28b4088a12ef84b Mon Sep 17 00:00:00 2001
From: Michal Simek <michal.simek@amd.com>
Date: Wed, 18 Mar 2026 15:24:52 +0100
Subject: [PATCH 46/47] efi_loader: Allow disabling ANSI console queries via
 Kconfig

Commit 4cb724364030 ("efi_loader: Disable ANSI output for tests")
introduced efi_console_set_ansi() to suppress ANSI escape sequences
during unit tests. Extend this mechanism to be configurable via a new
Kconfig option CONFIG_EFI_CONSOLE_DISABLE_ANSI.

When CONFIG_EFI_CONSOLE_DISABLE_ANSI is enabled,
efi_console_set_ansi(false) is called at the start of
efi_setup_console_size(). This prevents query_console_serial() from
sending ANSI escape sequences to the terminal, using default 25x80
dimensions instead. This is useful for platforms where the serial
console cannot handle ANSI queries.

Signed-off-by: Michal Simek <michal.simek@amd.com>
Reviewed-by: Ilias Apalodimas <ilias.apalodimas@linaro.org>
Reviewed-by: Heinrich Schuchardt <heinrich.schuchardt@canonical.com>
(cherry picked from commit 0c3eb097d9e827e21fdde7e6379f1a396db41da6)
---
 lib/efi_loader/Kconfig       | 11 +++++++++++
 lib/efi_loader/efi_console.c |  3 +++
 2 files changed, 14 insertions(+)

diff --git a/lib/efi_loader/Kconfig b/lib/efi_loader/Kconfig
index b7198108..3116a700 100644
--- a/lib/efi_loader/Kconfig
+++ b/lib/efi_loader/Kconfig
@@ -418,6 +418,17 @@ config EFI_EBBR_2_0_CONFORMANCE
 	help
 	  Enabling this option adds the EBBRv2.0 conformance entry to the ECPT UEFI table.
 
+config EFI_CONSOLE_DISABLE_ANSI
+	bool "Disable ANSI escape sequence queries for console size"
+	help
+	  Select this option to disable ANSI escape sequence queries for
+	  detecting serial console size. When enabled,
+	  efi_setup_console_size() will not send ANSI escape sequences to
+	  the terminal and will use default 25x80 dimensions instead.
+	  This is useful for platforms where the serial console cannot
+	  properly handle ANSI queries, causing boot delays or garbled
+	  output.
+
 config EFI_RISCV_BOOT_PROTOCOL
 	bool "RISCV_EFI_BOOT_PROTOCOL support"
 	default y
diff --git a/lib/efi_loader/efi_console.c b/lib/efi_loader/efi_console.c
index ecae86d0..18512a8c 100644
--- a/lib/efi_loader/efi_console.c
+++ b/lib/efi_loader/efi_console.c
@@ -356,6 +356,9 @@ void efi_setup_console_size(void)
 	int rows = 25, cols = 80;
 	int ret = -ENODEV;
 
+	if (IS_ENABLED(CONFIG_EFI_CONSOLE_DISABLE_ANSI))
+		efi_console_set_ansi(false);
+
 	if (IS_ENABLED(CONFIG_DM_VIDEO))
 		ret = query_vidconsole(&rows, &cols);
 	if (ret) {
-- 
2.55.0

